### ISSUE TITLE
`typeof this` stops working after `for..in` or `for..of`: "Cannot find name 'this'."

### ISSUE THREAD

**Issue by Yonom** [microsoft/TypeScript#46072]
# Bug Report

After a for..in or for..of statement, I can no longer access `typeof this` in class methods. A compile time error appears saying "Cannot find name 'this'."

### 🔎 Search Terms

Cannot find name 'this'. typeof this class for in for of

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about typeof, classes, typeof this

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.2#code/MYGwhgzhAEAqCmEAuMDeBYAUNH0mKQEYAKASmlWgHoroB5AaS1xbwE8AHeOA6AXnZcA9gDM8ACwCWEANzNcAXyzyc+ZACYyFarUYrWIoQCdoxGTPKol2Vqs7cEyfoPiiJ0uTZzX9apAGYtShpoADlGAC5oeAAPLmB8ABNoADshaGAhAFsOSRBueCMjYwh9FkMTYkyUp0SAVyystmhJFOgAbQBdS2tbOy4eJwEkezckKVl9Hy88AgAWIJ0wyOi4+AT4ZLSM7Nz86KKSstwK02rahqboNy6e4-6HXmHRsXGPKawFIA)

### 💻 Code

```ts
class Tests {
    test1() { // OK
        type Test = typeof this;
    }

    test2() { // OK
        for (;;) {}
        type Test = typeof this;
    }

    test3() { // NOK: expected no compile errors
        for (const dummy in []) {}
        type Test = typeof this; // Error 2304: Cannot find name 'this'.
    }

    test4() { // NOK: expected no compile errors
        for (const dummy of []) {}
        type Test = typeof this;  // Error 2304: Cannot find name 'this'.
    }
}
```

### 🙁 Actual behavior

The test cases `test3` and `test4` show compile errors.

### 🙂 Expected behavior

All test cases should pass without error.

**Comment by andrewbranch** [comment]
Congratulations, this is officially the weirdest bug I’ve seen today

![image](https://user-images.githubusercontent.com/3277153/135540185-15465fa5-b0ed-4b6a-a0ec-1d1638e6ef74.png)

Uh, also, all of these were errors until 4.4?!

**Comment by fatcerberus** [comment]
In the meantime, for the record: `typeof this` and `this` are the same type.

**Comment by paranoiacblack** [comment]
Something interesting to note here is that it compiles fine if there are any statements within the `for..in` or `for..of` loop.

**Comment by Zzzen** [comment]
`test2` is an infinite loop, so anything could happen.

**Comment by Zzzen** [comment]
>In the meantime, for the record: typeof this and this are the same type.

Actually, they are slightly different:  `this` always refers to the current scope of it while `typeof this` will use its  flow type.

```ts
class Base {
    f() {
        if (this instanceof D) {
            this.g();
            (this as this).g();  // Error: Property 'g' does not exist on type 'Base'.(2339)
            (this as typeof this).g();
        }
    }
}

class D {
    g() {}
}
```
